lib/services/data/providers/loki.js

```diff
@@ -0,0 +1,459 @@
+const db = require('lokijs'),
+  EventEmitter = require('events').EventEmitter,
+  fs = require('fs-extra'),
+  readline = require('readline'),
+  constants = require('../../../constants'),
+  async = require('async'),
+  util = require('../../utils/shared'),
+  _ = require('lodash'),
+  path = require('path');
+
+module.exports = class LokiDataProvider extends EventEmitter {
+  constructor(settings, logger) {
+    super();
+    this.settings = settings;
+    this.logger = logger;
+
+    this.initialize = util.maybePromisify(this.initialize);
+    this.stop = util.maybePromisify(this.stop);
+    this.upsert = util.maybePromisify(this.upsert);
+    this.merge = util.maybePromisify(this.merge);
+    this.insert = util.maybePromisify(this.insert);
+    this.find = util.maybePromisify(this.find);
+    this.findOne = util.maybePromisify(this.findOne);
+    this.remove = util.maybePromisify(this.remove);
+    this.count = util.maybePromisify(this.count);
+    this.operationCount = 0;
+  }
+  initialize(callback) {
+    this.dirty = true;
+    this.db = new db();
+    this.collection = this.db.addCollection('happn', {
+      indices: ['path', 'created', 'modified'],
+      unique: ['path']
+    });
+    this.persistenceOn = this.settings.filename != null;
+    this.operationQueue = async.queue((operation, cb) => {
+      this.processOperation(operation, cb);
+    }, 1);
+    if (!this.persistenceOn) {
+      callback();
+      return;
+    }
+    this.settings.snapshotRollOverThreshold = this.settings.snapshotRollOverThreshold || 1e3; // take a snapshot and compact every 1000 records
+    this.reconstruct(e => {
+      if (e) {
+        this.logger.error('failed reconstructing database', e);
+        callback(e);
+        return;
+      }
+      callback();
+    });
+  }
+  reconstruct(callback) {
+    if (!fs.existsSync(this.settings.filename)) {
+      return this.snapshot(callback);
+    }
+    const reader = readline.createInterface({
+      input: fs.createReadStream(this.settings.filename),
+      crlfDelay: Infinity
+    });
+    let lineIndex = 0;
+
+    reader.on('line', line => {
+      if (lineIndex === 0) {
+        this.db.loadJSON(JSON.parse(line).snapshot, { retainDirtyFlags: false });
+        this.collection = this.db.collections[0];
+      } else {
+        this.mutateDatabase(this.parsePersistedOperation(line));
+      }
+      lineIndex++;
+    });
+
+    reader.on('close', () => {
+      this.snapshot(callback);
+    });
+
+    reader.on('error', e => {
```

Comment:
are the 'close' and 'error' events mutually exclusive? worried about possibility of double callback